### DISCUSSION TITLE
Log thread CPU usage

### DISCUSSION THREAD

**Discussion by Steph71** [winsiderss/systeminformer#2035] [General Discussion]
Hi, I wonder whether there is any possibility to log a thread cpu usage over a period of time.

Any hints appreciated.

**Comment by dmex** [comment]
`typeperf "\thread(msedge/14)\% processor time"`

**Comment by Steph71** [comment]
Hi, this is definitely leading into the right direction. But it does not (yet) get me there. 

Let me elaborate a bit deeper:

I am using a simple python script, which is using threading module to create 2 threads in 1 process like this:

```py
from threading import Thread
from tqdm import tqdm
import os

def worker():
    x = 1.564243525424542
    for i in tqdm(range(10000000000), mininterval=1):
        x =x +10
        x = x / 10

if __name__ == "__main__":
    t1 = Thread(target=worker)
    t2 = Thread(target=worker)
    print(os.getpid())
    t1.start()
    t2.start()
```

Systeminformer is able to read cpu usage of these threads in a reasonable manner:

![image](https://github.com/winsiderss/systeminformer/assets/8713992/ee6f36b2-bd09-4cb6-b6ae-27860e14a0e6)

as python can use only 1 logical processor / process and my machine has 16 of them it makes perfect sense to have ~3% per thread reported.

I played with typeperf in order to retrieve the same information:

I can receive information of a specific process (given I know the PID) with

```
typeperf "\\ISCNPF4GFCGE\process v2(python:23008)\Thread Count"

"(PDH-CSV 4.0)","\\ISCNPF4GFCGE\process v2(python:23008)\Thread Count"
"04/22/2024 09:43:26.925","4.000000"
```

something like this:

```
typeperf -qx "\thread" | find "python" | find "ID"
\thread(python/0)\ID Process
\thread(python/0)\ID Process
\thread(python/0)\ID Process
\thread(python/1)\ID Process
\thread(python/2)\ID Process
\thread(python/3)\ID Process
...
\thread(python/0)\ID Process
\thread(python/0)\ID Process
\thread(python/1)\ID Process
\thread(python/2)\ID Process
...
```

will give me a list of python threads ... but the performance counters are not unique ... basically every python process has a "\thread(python/0)"

and reading the counter will of cause always give me the same result (probably the first it finds)

```
typeperf "\thread(python/0)\ID Process"

"(PDH-CSV 4.0)","\\ISCNPF4GFCGE\thread(python/0)\ID Process"
"04/22/2024 10:50:30.424","16340.000000"
"04/22/2024 10:50:31.439","16340.000000"
```

trying to provide the process id in the performance counter string does also not work unfortunately:

```
typeperf "\Thread(python:23512/0)\ID Process"

"(PDH-CSV 4.0)"
"04/22/2024 11:13:22.413","-1"
"04/22/2024 11:13:23.425","-1"
"04/22/2024 11:13:24.438","-1"
"04/22/2024 11:13:25.453","-1"
...
```

So I still have now way to get performance values of threads if the process name is not unique.

Obviously for this toy example I could rename my python execuatble and then iterate through the treads ... but this is not possible for my real usecase ...

additionally the tool is giving me quite shaky numbers in the cases it works:

```
typeperf  "\\ISCNPF4GFCGE\process v2(python:2080)\% Processor Time"

"(PDH-CSV 4.0)","\\ISCNPF4GFCGE\process v2(python:2080)\% Processor Time"
"04/22/2024 11:23:57.243","22.949787"
"04/22/2024 11:23:58.257","35.442335"
"04/22/2024 11:23:59.272","30.818078"
"04/22/2024 11:24:00.287","15.380030"
"04/22/2024 11:24:01.300","26.258965"
"04/22/2024 11:24:02.309","23.228927"
"04/22/2024 11:24:03.323","35.431831"
"04/22/2024 11:24:04.338","18.469211"
"04/22/2024 11:24:05.347","35.624754"
"04/22/2024 11:24:06.360","35.465315"
"04/22/2024 11:24:07.375","16.940178"
"04/22/2024 11:24:08.388","33.931455"```
```

even though the process I am observing has a very static load, basically it is using one logical processor for 100% of the time. Which can be nicely observed in system informer.

Not sure why it is dangling around 30 % as it should actually be at 6% (1/16) statically.

Don't get confused about the changing PIDs in the thread  ... I started the process a couple of times during my experiments ... general statements apply.

Any comments an suggestions appreciated.

Best Regards
Steph

**Comment by Steph71** [comment]
further looking into this issue I found 2 more tools, that basically do the same thing:
1.) [Sysinternals Process explorer](https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer)

unfortunately this gives as unreasonable process an thread cpu utilization readings as typeperf does.

2.) Then I stumbled across a power [shell script](https://xkln.net/blog/analyzing-thread-cpu-utilization-with-processexplorer-powershell-and-wmi/)

but also here cpu usage readings don't make sense to me.

Seems like systeminformer is doing something right what everybody else is doing wrong ... I would love to have a option to log this infos over time ...

![image](https://github.com/winsiderss/systeminformer/assets/8713992/d0f4e857-ea6d-4714-a104-6893b2f1a727)

**Comment by Steph71** [comment]
interesting, same experiment as above but with an older PC (i7-2600) running Windows 10. Everything as expected ...
... probably the tools cannot handle the different core types (P vs E) the 12th gen Intel CPU.
... or its just a windows 11 quirk ...
![image](https://github.com/winsiderss/systeminformer/assets/8713992/d5b40401-8474-473d-a32b-a646488b472b)
